ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/FilterStringColLikeStringScalar.java

```diff
@@ -140,4 +133,84 @@ public Checker tryCreate(String pattern) {
       return new ComplexChecker("^" + UDFLike.likePatternToRegExp(pattern) + "$");
     }
   }
+
+  /**
+   * Check the LIKE pattern..
+   */
+  private static class CheckerPattern {
+    private final String pattern;
+    private final UDFLike.PatternType type;
+
+    public CheckerPattern(UDFLike.PatternType type, String pattern) {
+      this.type = type;
+      this.pattern = pattern;
+    }
+
+    public String check() {
+      UDFLike.PatternType lastType = UDFLike.PatternType.NONE;
+      int length = pattern.length();
+      int beginIndex = 0;
+      int endIndex = length;
+      char lastChar = 0;
+      String strPattern = new String();
+      String simplePattern;
+
+      for (int i = 0; i < length; i++) {
```

Comment:
@ryukobayashi, are you saying that your code with char matching is more efficient than regex?  
Could you please create a benchmark (JMH) and share the results?

Reply:
@deniskuzZ  This was originally a bug in `java.util.regex`, so it was a fix to use `com.google.re2j` instead of using regex as is. The comments below led me to implement a custom algorithm: https://github.com/apache/hive/pull/3117#issuecomment-1076183992

So, I don't have any particular attachment to custom algorithms. Which is better?

Reply:
👍, do we have any benchmarks that prove efficiency of implemented algorithm? also, usually for such kind of tasks (pattern matching) state machines are one of the recommended techniques.

Reply:
@deniskuzZ Sorry for the late reply as I was unwell. This algorithm is not something I directly implemented. It's already implemented in a UDF that runs when Vectorized is off. So I don't know if there are any benchmarks to prove its efficiency: https://github.com/apache/hive/blob/release-4.0.0-rc0/ql/src/java/org/apache/hadoop/hive/ql/udf/UDFLike.java#L108-L158

Reply:
oh ok, please give me some time to check

Reply:
shouldn't `parseSimplePattern` logic be the same? if yes, could we move this to util class and reuse in both places?

Reply:
I will check and try this.